src/hooks/useSwapTaxes.ts

```diff
@@ -0,0 +1,82 @@
+import { InterfaceEventName } from '@uniswap/analytics-events'
+import { ChainId, Percent } from '@uniswap/sdk-core'
+import { WETH_ADDRESS as getWethAddress } from '@uniswap/universal-router-sdk'
+import { useWeb3React } from '@web3-react/core'
+import FOT_DETECTOR_ABI from 'abis/fee-on-transfer-detector.json'
+import { FeeOnTransferDetector } from 'abis/types'
+import { sendAnalyticsEvent } from 'analytics'
+import { ZERO_PERCENT } from 'constants/misc'
+import { useEffect, useState } from 'react'
+
+import { useContract } from './useContract'
+
+const FEE_ON_TRANSFER_DETECTOR_ADDRESS = '0x19C97dc2a25845C7f9d1d519c8C2d4809c58b43f'
+
+function useFeeOnTransferDetectorContract(): FeeOnTransferDetector | null {
+  const { account } = useWeb3React()
+  const contract = useContract<FeeOnTransferDetector>(FEE_ON_TRANSFER_DETECTOR_ADDRESS, FOT_DETECTOR_ABI)
+
+  useEffect(() => {
+    if (contract && account) {
+      sendAnalyticsEvent(InterfaceEventName.WALLET_PROVIDER_USED, {
+        source: 'useFeeOnTransferDetectorContract',
+        contract,
+      })
+    }
+  }, [account, contract])
+  return contract
+}
+
+// TODO(WEB-2787): add tax-fetching for other chains
+const WETH_ADDRESS = getWethAddress(ChainId.MAINNET)
+const AMOUNT_TO_BORROW = 10000 // smallest amount that has full precision over bps
+
+const FEE_CACHE: { [address in string]?: { sellTax?: Percent; buyTax?: Percent } } = {}
```

Comment:
I like the idea of a cache, but I'm wondering if the cost / benefit analysis works out in practice. Specifically, there may be a risk with staleness, since it's possible for a token's tax rates to change, in which case the cache would return the wrong values. Also, since the call is made through the user's provider, what is the benefit? Is there a high latency on using the FOT detector? Perhaps the increased speed would be a worthwhile value prop.

I definitely see caching being valuable when FOT detection is done in our backend (i.e. impacting latency of routing API) or if we're using our own providers and will incur the cost of these calls (as is the case for multichain features). 

What do you think?

Reply:
We're still using our own provider a good portion of the time, e.g. if the user hasn't connected a wallet yet but they're setting up swaps, or if they're connected via WalletConnect.

Staleness is definitely possible here, but uncommon to happen during the lifetime of a swap. I think it's best to conserve resources / performance rather than support tokens that change rates super quickly, and rely on the BE to handle that in the future if it becomes a problem. wdyt?

Reply:
I agree that staleness won't be common and definitely think we should prioritize resource conservation - thanks for clarifying those cases in which we're actually using our own provider!